### PR TITLE
TF: CTRL with native embedding layers

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Follows up on #23436, migrating the embedding layer of CTRL to native Keras embeddings

CTRL needed several related changes, so it deserves a stand-alone PR. This PR:
1. Replaces `TFSharedEmbeddings` by the native Keras layers
2. Fixes resized bias serialization, just like https://github.com/huggingface/transformers/pull/19013 does for BART -- in the process, gets rid of the separate LMHead class, which is outdated and tied to code scheduled for deprecation, and moves functions like `set_bias` to the right place
3. Fixes XLA issues (`prepare_inputs_for_generation` was incomplete)

### CODE REVIEW

File: tests/test_modeling_tf_common.py [hunk 1]
```diff
@@ -1042,6 +1042,7 @@ def test_model_common_attributes(self):
 
         for model_class in self.all_model_classes:
             model = model_class(config)
+            model(model.dummy_inputs)  # build the model
```
Comment (gante) [inline tests/test_modeling_tf_common.py:1045]: needed for the `get_bias()` call below --` model.bias_layer.bias` doesn't exist unless the model gets built

File: tests/models/ctrl/test_modeling_tf_ctrl.py [hunk 2]
```diff
@@ -216,33 +216,6 @@ def test_ctrl_sequence_classification_model(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_ctrl_for_sequence_classification(*config_and_inputs)
 
-    def test_model_common_attributes(self):
-        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
-        list_lm_models = [TFCTRLLMHeadModel]
-        list_other_models_with_output_ebd = [TFCTRLForSequenceClassification]
```
Comment (gante) [inline tests/models/ctrl/test_modeling_tf_ctrl.py:222]: `TFCTRLForSequenceClassification` incorrectly had a `get_output_embeddings()`, and got the test overridden -- sequence classification models do not have a LM Head (and therefore no output embeddings)

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 3]
```diff
@@ -305,7 +305,7 @@ def call(
             # So we can broadcast to [batch_size, num_heads, from_seq_length, to_seq_length]
             # this attention mask is more simple than the triangular masking of causal attention
             # used in OpenAI GPT, we just need to prepare the broadcast dimension here.
-            attention_mask = tf.reshape(attention_mask, (input_shape[0], 1, 1, input_shape[1]))
+            attention_mask = tf.reshape(attention_mask, (input_shape[0], 1, 1, input_shape[1] + past_length))
```
Comment (gante) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:308]: `+ past_length` here accounts for generation with caching, where the attention mask has the length of `input_ids` (1 token) + all past tokens

It is also part of the XLA fixes

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 4]
```diff
@@ -571,39 +571,26 @@ def serving_output(self, output):
         )
 
 
-class TFCTRLLMHead(tf.keras.layers.Layer):
-    def __init__(self, config, input_embeddings, **kwargs):
-        super().__init__(**kwargs)
-        self.config = config
-        # CTRL has numerical issues in XLA generate
-        self.supports_xla_generation = False
+class BiasLayer(tf.keras.layers.Layer):
```
Comment (gante) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:574]: Slightly different than the separate LM bias head for BART, because of the serialization name. In BART, the bias weights are at root level (and thus added in `__init__`). Here, we want it nested as `lm_head/bias:0`, so it gets initialized in `build()`.

File: src/transformers/models/ctrl/modeling_tf_ctrl.py [hunk 5]
```diff
@@ -737,9 +750,6 @@ def __init__(self, config, *inputs, **kwargs):
         )
         self.transformer = TFCTRLMainLayer(config, name="transformer")
 
-    def get_output_embeddings(self):
```
Comment (gante) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:740]: `TFCTRLForSequenceClassification` incorrectly had a `get_output_embeddings()`, and got the test overridden -- sequence classification models do not have a LM Head (and therefore no output embeddings)

I also removed the overridden test

**Review by Rocketknight1 — APPROVED** [review]
Looks like another good cleanup to me!

File: tests/test_modeling_tf_common.py (hunk 1 above)
Comment (amyeroberts) [reply tests/test_modeling_tf_common.py:1045]: `needed for the get_bias() call below ` Would be a better comment that `build the model` :)

File: src/transformers/models/ctrl/modeling_tf_ctrl.py (hunk 4 above)
Comment (amyeroberts) [inline src/transformers/models/ctrl/modeling_tf_ctrl.py:574]: Should still have the model name as a prefix

```suggestion
class TFCTRLBiasLayer(tf.keras.layers.Layer):
```

File: src/transformers/models/ctrl/modeling_tf_ctrl.py (hunk 5 above)
Comment (amyeroberts) [reply src/transformers/models/ctrl/modeling_tf_ctrl.py:740]: Not so fast! :D 

Even if it's not correct to have this method - it's one that belongs to a public model class. We should have a deprecation cycle first warning users that it will be deleted in a few versions' time

**Review by amyeroberts — COMMENTED** [review]
Thanks for this tidy up! 

Overall looks good - main comment is about not removing public methods

**Comment by gante** [comment]
@amyeroberts comments addressed 🙌

**Review by amyeroberts — APPROVED** [review]
Thanks for updating!
